woocommerce/woocommerce-gutenberg-products-block
# Single Product Block

Creates the basic structure for the Single Product Block (behind the experimental flag).

Closes #2392
Closes #2535

- Creates the block, with icon and description
- Implements the product picker in edit mode and the inspector
- Moves the inner block layout functionality to shared atomic utils (used by all products and single product blocks)
  - utils also include block->component maps and anything else related to atomic components
- ~~Introduces a new hook named `useSyncedLayoutConfig ` which syncs inner block changes to block attributes as a LayoutConfig array. This is only being used by Single Products Block currently. The All Products block does not update instantly and so does not need this (yet)~~
- **Introduces a new method of rendering inner blocks without layout configs.** (https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2542)
- Removes dedicated atomic/components dir so that the components can share attribute data with atomic/blocks when rendering on the frontend.
- Combines `useProductLayoutContext` and `useInnerBlockConfigurationContext` since both are used at the same time in all contexts. This allows the parent block to be passed, as well as a prefix for the CSS classnames.
- Introduces `ProductDataContext` to share real product data with atomic blocks.
   - For all products block, this is where _preview data_ is passed. Prior to this change, the atomic blocks themselves looked up preview data.
   - For the single product block, it passes real product data via this context.
   - Atomic blocks feed this data to the atomic components
- Updates atomic blocks to have proper loading states and consume `ProductDataContext`.
   - Prior to this some of the atomic blocks had loading states, but others did not. I've improved this so anytime _product_ is null, it's considered in loading state. This is useful if the product data is being pulled from the API.
- Shared contexts so Atomic blocks are functional (see https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2399#issuecomment-631560641) This ensures contexts work when wrapping atomic blocks in the editor and on frontend.

_Known follow ups:_
- Style duplication will be resolved in #2527. Right now there are duplicate styles across all products and single products which could be improved.
- The preview content will be added in a follow up when the design is finalised.
- The inner block editor will be further developed as part of #2394
- There is a basic atomic block structure in this PR as an example, but it won't be completed until #2393 is resolved.
- There is a basic implementation of columns support as a POC but this will be looked at some more in #2526

**Screenshots**

Edit interface:
![Screenshot 2020-05-06 at 10 14 14](https://user-images.githubusercontent.com/90977/81160850-877f2a80-8f82-11ea-8aa6-fff5448a6710.png)
![Screenshot 2020-05-06 at 10 15 26](https://user-images.githubusercontent.com/90977/81160846-85b56700-8f82-11ea-9064-e8fe7e0b6841.png)

Loading state:
![placeholder](https://user-images.githubusercontent.com/90977/82566558-0031d980-9b74-11ea-9a08-7676393fffcb.png)

Inner Block state:
![content](https://user-images.githubusercontent.com/90977/82566573-0627ba80-9b74-11ea-8987-668985cafeba.png)

**How to test**

1. Confirm _All Products Block_ remains functional in editor/frontend
2. In development mode, confirm the block is built and shown in the editor
3. Insert the block and select a product. Ensure it persists on save.
4. Try toggling between edit and layout mode.
5. Confirm placeholder details are shown on frontend.
